netaddr.go

```diff
@@ -252,6 +252,75 @@ func (ip IP) IsMulticast() bool {
 	}
 }
 
+// Mask applies a CIDR mask of leading bits to IP, producing an IPPrefix
+// of the specified length. If IP is the zero value or bits is too large, a
+// zero-value IPPrefix is returned.
+func (ip IP) Mask(bits uint8) IPPrefix {
+	maxBits := uint8(32)
+	if ip.Is6() {
+		maxBits = 128
+	}
+
+	if bits > maxBits {
+		// Invalid mask size for both IPv4/6: nothing to do.
+		return IPPrefix{}
+	}
+
+	var b []byte
+	switch ip := ip.ipImpl.(type) {
+	case nil:
+		// Zero-value input produces zero-value output.
+		return IPPrefix{}
+	case v4Addr:
+		b = ip[:]
+	case v6Addr:
+		b = ip[:]
+	case v6AddrZone:
+		b = ip.v6Addr[:]
+	default:
+		panic("netaddr: unhandled ipImpl representation")
+	}
+
+	// Apply the mask specified by bits.
+	n := bits
+	for i := 0; i < len(b); i++ {
+		if n >= 8 {
+			b[i] &= 0xff
+			n -= 8
+			continue
+		}
+
+		b[i] = ^byte(0xff >> n)
+		n = 0
+	}
+
+	var out IP
+	switch ip.ipImpl.(type) {
+	case v4Addr:
+		var v4 v4Addr
+		copy(v4[:], b)
+		out = IP{ipImpl: v4}
+	case v6Addr:
+		var v6 v6Addr
+		copy(v6[:], b)
+		out = IP{ipImpl: v6}
+	case v6AddrZone:
+		// TODO: does it makes sense to apply zone in an IPPrefix?
```

Comment:
It's unusual, but I don't see a reason to make it not-work. If only so that you don't have to treat zoned v6 addrs specially in generic code, unless you specifically care.

Reply:
I'm less bothered here because the invalid input is unlikely to come from users/network. It's just a number.

But I could go either way.